File: src/lib/placementOrdering.ts
import type { IsoDate, Placement } from "@/types";

/**
 * Pure placement-ordering algorithm (ADR-0015 §2), shared by the DB layer (src/lib/db/index.ts)
 * and the store's optimistic client-side patches (src/store/tripStore.ts) so both apply the exact
 * same reordering semantics — no risk of the client's optimistic guess drifting from what the
 * server actually persists.
 */

/** Shift siblings on `date` at/after `order` up by one, opening a slot. */
function shiftInto(placements: Placement[], date: IsoDate, order: number): Placement[] {
  return placements.map((p) => (p.date === date && p.order >= order ? { ...p, order: p.order + 1 } : p));
}

/** Re-densify `date`'s placements to a gap-free 0..n-1 order sequence. */
function densify(placements: Placement[], date: IsoDate): Placement[] {
  const onDate = [...placements].filter((p) => p.date === date).sort((a, b) => a.order - b.order);
  const orderById = new Map(onDate.map((p, i) => [p.id, i]));
  return placements.map((p) => (orderById.has(p.id) ? { ...p, order: orderById.get(p.id)! } : p));
}

/**
 * Move an existing placement to `date`/`order`. Siblings at/after the target order shift down; if
 * the placement left another date, that date's remaining placements are re-densified.
 */
export function reorderPlacements(
  placements: Placement[],
  placementId: string,
  date: IsoDate,
  order: number
): Placement[] {
  const current = placements.find((p) => p.id === placementId);
  if (!current) throw new Error("Placement not found");
  const sourceDate = current.date;

  let next = shiftInto(placements, date, order);
  next = next.map((p) => (p.id === placementId ? { ...p, date, order } : p));
  if (sourceDate !== date) next = densify(next, sourceDate);
  return next;
}

/**
 * Insert a new placement for `locationId` on `date`. Appends to the end of the date unless `order`
 * is given, in which case siblings at/after it shift down to make room.
 */
export function insertPlacement(
  placements: Placement[],
  tripId: string,
  newPlacement: { id: string; locationId: string; date: IsoDate; order?: number }
): Placement[] {
  const { id, locationId, date } = newPlacement;
  let order = newPlacement.order;
  let next = placements;
  if (order === undefined) {
    order = placements.filter((p) => p.date === date).reduce((max, p) => Math.max(max, p.order), -1) + 1;
  } else {
    next = shiftInto(placements, date, order);
  }
  return [...next, { id, tripId, locationId, date, order }];
}

File: swift/TripKrakenKit/Sources/TripKrakenKit/PlacementOrdering.swift
/// Pure placement-ordering algorithm (ADR-0015 §2), shared by the persistence layer and any
/// optimistic client-side patch so both apply the exact same reordering semantics — no risk of a
/// client's optimistic guess drifting from what actually gets persisted.

public enum PlacementOrderingError: Error, Sendable {
    case placementNotFound
}

/// Shift siblings on `date` at/after `order` up by one, opening a slot.
private func shiftInto(_ placements: [Placement], date: IsoDate, order: Int) -> [Placement] {
    placements.map { p in
        guard p.date == date, p.order >= order else { return p }
        var shifted = p
        shifted.order += 1
        return shifted
    }
}

/// Re-densify `date`'s placements to a gap-free 0..n-1 order sequence.
private func densify(_ placements: [Placement], date: IsoDate) -> [Placement] {
    let onDate = placements.filter { $0.date == date }.sorted { $0.order < $1.order }
    let orderById = Dictionary(uniqueKeysWithValues: onDate.enumerated().map { ($1.id, $0) })
    return placements.map { p in
        guard let order = orderById[p.id] else { return p }
        var densified = p
        densified.order = order
        return densified
    }
}

/// Move an existing placement to `date`/`order`. Siblings at/after the target order shift down; if
/// the placement left another date, that date's remaining placements are re-densified.
public func reorderPlacements(
    _ placements: [Placement], placementId: String, date: IsoDate, order: Int
) throws(PlacementOrderingError) -> [Placement] {
    guard let current = placements.first(where: { $0.id == placementId }) else {
        throw .placementNotFound
    }
    let sourceDate = current.date

    var next = shiftInto(placements, date: date, order: order)
    next = next.map { p in
        guard p.id == placementId else { return p }
        var moved = p
        moved.date = date
        moved.order = order
        return moved
    }
    if sourceDate != date { next = densify(next, date: sourceDate) }
    return next
}

/// Insert a new placement for `locationId` on `date`. Appends to the end of the date unless `order`
/// is given, in which case siblings at/after it shift down to make room.
public func insertPlacement(
    _ placements: [Placement], id: String, tripId: String, locationId: String, date: IsoDate,
    order: Int? = nil
) -> [Placement] {
    var next = placements
    let resolvedOrder: Int
    if let order {
        resolvedOrder = order
        next = shiftInto(placements, date: date, order: order)
    } else {
        resolvedOrder = placements.filter { $0.date == date }.reduce(-1) { max($0, $1.order) } + 1
    }
    return next + [Placement(id: id, tripId: tripId, locationId: locationId, date: date, order: resolvedOrder)]
}
